libraries/state_history/include/eosio/state_history/log.hpp

```diff
@@ -261,16 +332,103 @@ class state_history_log {
       return log;
    }
 
+   /// @return the decompressed entry size
+   uint64_t get_unpacked_entry(uint32_t block_num, std::variant<std::vector<char>, maybe_locked_compress_stream>& result) {
+      if (block_num < _begin_block || block_num >= _end_block)
+         return 0;
+
+      state_history_log_header header;
+      std::lock_guard lock(mx);
+      log.seek(get_pos(block_num));
+      read_header(header);
+
+      uint64_t payload_size = header.payload_size;
+      uint32_t s;
+      uint64_t compressed_size;
+
+      log.read((char*)&s, sizeof(s));
+      if (s == 1 && payload_size > (s + sizeof(uint32_t))) {
+         compressed_size = payload_size - sizeof(uint32_t) - sizeof(uint64_t);
+         uint64_t decompressed_size = 0;
+         log.read((char*)&decompressed_size, sizeof(decompressed_size));
+
+         EOS_ASSERT(decompressed_size < 0x8000000000000000, chain::plugin_exception, "error reading decompressed_size=${decompressed_size},  file = ${fn}, pos=${pos}", ("decompressed_size", decompressed_size)("fn", log.get_file_path().string())("pos", log.tellp()));
```

Comment:
I think this would be a place we should use `FC_ASSERT` instead of `EOS_ASSERT`.